src/core/or/circuitpadding.h

```diff
@@ -0,0 +1,539 @@
+/*
+ * Copyright (c) 2017, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * \file circuitpadding.h
+ * \brief Header file for circuitpadding.c.
+ **/
+#ifndef TOR_CIRCUITPADDING_H
+#define TOR_CIRCUITPADDING_H
+
+#include "circpad_negotiation.h"
+#include "lib/evloop/timers.h"
+
+typedef struct circuit_t circuit_t;
+typedef struct origin_circuit_t origin_circuit_t;
+typedef struct cell_t cell_t;
+
+/**
+ * Circpad state specifier.
+ *
+ * Each circuit has up to two state machines, and each state
```

Comment:
I had thought that it had up to one state machine for each hop?

Reply:
No, each circuit has up to two machines. They can be both pad to the same hop (which is what WTF-PAD and APE do -- they use one machine for handling received packets and a second machine for handling sent packets), or they can be at different hops.